client/adventureMap/CInfoBar.cpp

```diff
@@ -163,16 +163,62 @@ CInfoBar::VisibleGameStatusInfo::VisibleGameStatusInfo()
 	}
 }
 
-CInfoBar::VisibleComponentInfo::VisibleComponentInfo(const Component & compToDisplay, std::string message)
+CInfoBar::VisibleComponentInfo::VisibleComponentInfo(const std::vector<Component> & compsToDisplay, std::string message, int textH, bool tiny)
 {
 	OBJECT_CONSTRUCTION_CAPTURING(255-DISPOSE);
 
 	background = std::make_shared<CPicture>("ADSTATOT", 1, 0);
+	auto fullRect = Rect(0, 0, data_width, data_height);
+	auto textRect = fullRect;
+	auto imageRect = fullRect;
+	auto font = FONT_SMALL;
 
-	comp = std::make_shared<CComponent>(compToDisplay);
-	comp->moveTo(Point(pos.x+47, pos.y+50));
+	if(!compsToDisplay.empty())
+	{
+		auto size = CComponent::large;
+		if(compsToDisplay.size() > 2)
+			size = CComponent::small;
+		if(!message.empty())
+		{
+			textRect = Rect(CInfoBar::component_offset/2, 
+							CInfoBar::component_offset/2,
+							data_width - CInfoBar::component_offset,
+							textH);
+			imageRect = Rect(CInfoBar::component_offset/2,
+							 textH + CInfoBar::component_offset/2,
+							 data_width - CInfoBar::component_offset,
+							 CInfoBar::data_height - CInfoBar::component_offset - textH);
+			if(compsToDisplay.size() > 4)
+				size = CComponent::tiny;
+		}
+		else if(compsToDisplay.size() > 4)
+			size = CComponent::small;
+
+		std::vector<std::shared_ptr<CComponent>> vect;
+
+		for(const auto & c : compsToDisplay)
+			vect.emplace_back(std::make_shared<CComponent>(c, size));
+
+		comps = std::make_shared<CComponentBox>(vect, imageRect, 4, 4, 1);
+	}
+	else
+		font = tiny ? FONT_TINY : font;
 
-	text = std::make_shared<CTextBox>(message, Rect(10, 4, 160, 50), 0, FONT_SMALL, ETextAlignment::CENTER, Colors::WHITE);
+	if(!message.empty())
+		text = std::make_shared<CTextBox>(message, textRect, 0, font, ETextAlignment::CENTER, Colors::WHITE);
+}
+
+int CInfoBar::getEstimatedComponentHeight(int numComps) const
+{
+	if (numComps > 8) //Bigger than 8 components - return invalid value
+		return std::numeric_limits<int>::max();
```

Comment:
```suggestion
	assert(numComps <= 8);
	if (numComps > 8) //Bigger than 8 components - return invalid value
		return std::numeric_limits<int>::max();
```

Reply:
@IvanSavenko assert is not good here, because modder can set 9 components as reward, and then we just return false from infowindow, and show modal window instead.